src/Runner.Worker/CompositeStepsRunner.cs

```diff
@@ -0,0 +1,249 @@
+                }
+
+                // Stomps over with outside step env
+                if (step.ExecutionContext.ExpressionValues.TryGetValue("env", out var envContextData))
+                {
+#if OS_WINDOWS
+                    var dict = envContextData as DictionaryContextData;
+#else
+                    var dict = envContextData as CaseSensitiveDictionaryContextData;
+#endif
+                    foreach (var pair in dict)
+                    {
+                        envContext[pair.Key] = pair.Value;
+                    }
+                }
+
+                step.ExecutionContext.ExpressionValues["env"] = envContext;
+
+
+
+                if (step is IActionRunner actionStep)
+                {
+                    // Set GITHUB_ACTION
+                    step.ExecutionContext.SetGitHubContext("action", actionStep.Action.Name);
+
+                    try
+                    {
+                        // Evaluate and merge action's env block to env context
+                        var templateEvaluator = step.ExecutionContext.ToPipelineTemplateEvaluator();
+                        var actionEnvironment = templateEvaluator.EvaluateStepEnvironment(actionStep.Action.Environment, step.ExecutionContext.ExpressionValues, step.ExecutionContext.ExpressionFunctions, VarUtil.EnvironmentVariableKeyComparer);
+                        foreach (var env in actionEnvironment)
+                        {
+                            envContext[env.Key] = new StringContextData(env.Value ?? string.Empty);
+                        }
+                    }
+                    catch (Exception ex)
+                    {
+                        // fail the step since there is an evaluate error.
+                        Trace.Info("Caught exception in Composite Steps Runner from expression for step.env");
+                        // evaluateStepEnvFailed = true;
+                        step.ExecutionContext.Error(ex);
+                        // CompleteStep(step, TaskResult.Failed);
+                    }
+                }
+
+                // We don't have to worry about the cancellation token stuff because that's handled by the composite action level (in the StepsRunner)
+
+                await RunStepAsync(step);
+
+                // TODO: Add compat for other types of steps.
+            }
+            // Completion Status handled by StepsRunner for the whole Composite Action Step
+        }
+
+        private async Task RunStepAsync(IStep step)
+        {
+            // Check to see if we can expand the display name
+            if (step is IActionRunner actionRunner &&
+                actionRunner.Stage == ActionRunStage.Main &&
+                actionRunner.TryEvaluateDisplayName(step.ExecutionContext.ExpressionValues, step.ExecutionContext))
+            {
+                step.ExecutionContext.UpdateTimelineRecordDisplayName(actionRunner.DisplayName);
```

Comment:
you don't want to update the displayname for the record.

```
if (step is IActionRunner actionRunner && actionRunner.Stage == ActionRunStage.Main)
{
 actionRunner.TryEvaluateDisplayName(step.ExecutionContext.ExpressionValues, step.ExecutionContext);
}
```

Reply:
@TingluoHuang 

If I remove that, the UI now looks like: https://drive.google.com/file/d/1JzAWi0qOHQSOK-fBMp8Q_AFwPkkRqvDl/view?usp=sharing

But having the old code actually creates a more intuitive UI based on my conversations with Chris: https://drive.google.com/file/d/1-NrI8M1SoMKwS0y7LAl3JgmafzeVkapU/view?usp=sharing

Reply:
For now, I'll keep it as is